iwyu-check-license-header.py

```diff
@@ -0,0 +1,266 @@
+    return r.splitlines()
+
+
+def make_license_header(filename, one, two):
+    """ Build a valid license header from filename and comment styles.
+
+    The one and two args should be '#' and '##' or '//' and '//', respectively.
+    Returns a list of lines.
+    """
+    assert len(two) == 2
+    filename = os.path.basename(filename)
+
+    def dashes():
+        c = 80
+        c -= 2  # take off leading ##
+        c -= 6  # take off leading ===---
+        c -= len(filename) + 2  # take off filename with a space on either side
+        c -= 3  # take off trailing ===
+        c -= 2  # take off trailing ##
+        return '-' * c
+
+    firstline = '%s===--- %s %s===%s' % (two, filename, dashes(), two)
+    return [firstline] + make_hdrformat(one, two)
+
+
+def format_file_error(filename, *lines):
+    """ Format an error message from filename and lines """
+    lines = list(lines)
+    lines[0] = '%s: %s' % (filename, lines[0])
+    return os.linesep.join(lines)
+
+
+def find_license_header(lines, two):
+    """ Return an index where the license header begins """
+    if not lines:
+        return -1
+
+    for i, line in enumerate(lines):
+        # Allow leading blank lines and hash-bangs.
+        if not line or line.startswith('#!'):
+            continue
+
+        # Besides those, the first line should be a license header.
+        if line.startswith(two + '==='):
+            break
+
+        # If not, this fails the test entirely.
+        return -1
+
+    return i
+
+
+class File(object):
+    """ Base class for a source file with a license header
+
+    Do not use directly, instead use File.parse to instantiate a more derived
+    class.
+
+    Derived classes must have three class variables:
+
+    * one - One comment char ('#' for Python, '//' for C++)
+    * two - Two comment chars ('##' for Python, '//' for C++)
+    * pattern - a regex matching the first line in a license header
+    """
+    def __init__(self, filename):
+        with open(filename, 'r') as fd:
+            content = fd.read()
+
+        self.lines = list(content.splitlines())
+        self.hdrindex = find_license_header(self.lines, self.two)
+        self.filename = filename
+        self.errors = []
+
+    @classmethod
+    def parse(_, filename):
+        """ Return an object derived from File to analyze license headers """
+        _, ext = os.path.splitext(filename)
+        if ext == '.py':
+            klass = PythonFile
+        elif ext in ('.h', '.c', '.cc'):
+            klass = CxxFile
+        else:
+            return None
+
+        return klass(filename)
+
+    def has_license_header(self):
+        """ Return True if a license header has been found """
+        return self.hdrindex != -1
+
+    def add_license_header(self):
+        """ Add license header to a file that doesn't have one. """
+        assert not self.has_license_header()
+
+        # Find insertion point
+        for p, line in enumerate(self.lines):
```

Comment:
This construct is used twice. I think a function that returns the expected line where the license should start would be clearer. 
e.g.
```py
p = find_expected_license_start_line(lines) # -1 < p <= len(lines)
```

Reply:
I've toyed with this, but I have yet to find a meaningful name for that function that's as clear as the simple open-coded implementation.

In the one place, it's used to detect if there is a license header, and in the other to figure out where to put it, so the semantics are subtly different.

Reply:
I left this unchanged. Not too worried about it.